app/src/test/java/com/codesoom/assignment/controllers/TaskControllerNestedTest.java

```diff
@@ -0,0 +1,302 @@
+package com.codesoom.assignment.controllers;
+
+import com.codesoom.assignment.TaskNotFoundException;
+import com.codesoom.assignment.application.TaskService;
+import com.codesoom.assignment.models.Task;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class TaskControllerNestedTest {
+
+    private final String DEFAULT_TITLE = "TEST";
+    private final long DEFAULT_SIZE = 3;
+    private TaskController controller;
+
+    @BeforeEach
+    void setUp() {
+        // subject
+        controller = new TaskController(new TaskService());
+
+        // fixtures
+    }
+
+    void setDefaultSizeTasks(){
+        for(long i = 1 ; i <= DEFAULT_SIZE ; i++){
+            Task newTask = new Task();
+            newTask.setTitle(DEFAULT_TITLE + i);
+            controller.create(newTask);
+        }
+    }
+
+    @Nested
+    @DisplayName("list 메소드는")
+    class Describe_List{
+
+        @Nested
+        @DisplayName("Task가 존재한다면")
+        class Context_ExistsTask{
+
+            @BeforeEach
+            void setUp(){
+                setDefaultSizeTasks();
+            }
+
+            @Test
+            @DisplayName("모든 Task를 반환한다")
+            void It_ReturnAllTask(){
+                assertThat(controller.list()).hasSize((int) DEFAULT_SIZE);
+                for(long i = 1 ; i <= DEFAULT_SIZE ; i++){
+                    assertThat(controller.detail(i)).isEqualTo(new Task(i , DEFAULT_TITLE + i));
+                }
+            }
+        }
+
+        @Nested
+        @DisplayName("Task가 존재하지 않는다면")
+        class Context_NotExistsTask{
+            private final int EMPTY = 0;
+
+            @Test
+            @DisplayName("비어있는 List를 반환한다")
+            void It_ReturnEmptyList(){
+                assertThat(controller.list()).hasSize(EMPTY);
+            }
+        }
+
+    }
+
+    @Nested
+    @DisplayName("detail 메소드는")
+    class Describe_Detail{
+
+        @Nested
+        @DisplayName("id가 null이라면")
+        class Context_NullPathVariable{
+            private final Long givenId = null;
+
+            @Test
+            @DisplayName("TaskNotFoundException를 던진다")
```

Comment:
잘 수정하셨습니다. 그런데 여기에서 한 가지 더 생각해볼만한 문제가 있습니다. 바로 `TaskNotFoundException` 클래스의 이름 변경입니다. 누가 `Task` 클래스의 이름을 `Todo`로 변경한다고 상상해 봅시다. IntelliJ 에서 rename 해서 클래스 이름을 변경하겠죠? 그리고 `Task`가 `Todo`로 바뀌었으니까, 작업한 사람은 `TaskNotFoundException` 클래스가 있는 걸 보고 "이것도 바꿔야겠다"라고 생각하고 `TodoNotFoundException`으로 이름을 바꿀 거에요. 그리고 `Task`가 들어간 다른 클래스들도 몇 개 더 바꿨다고 상상해 보세요.

아마 그 분은 이 테스트의 `@DisplayName`의 문자열도 수정해야 한다는 걸 놓칠 수도 있을 겁니다. IntelliJ가 문자열까지 rename해주지는 않으니까요. 그래서 rename하는 사람이 이 설명을 수정하지 않고 넘어갈 수 있겠죠. 거짓된 설명이 남게 되는 것입니다. 

```suggestion
            @DisplayName("찾지 못했다는 내용의 예외를 던진다")
```

하지만 위와 같이 설명을 작성해 두면 클래스 이름 변경에 취약했던 설명이 이름 변경에도 유연하게 대응하게 됩니다. 즉 클래스 이름에 테스트 설명이 강하게 의존하고 있었던 문제를 약하게 의존하게 만들어 해소한거죠. 잘 생각해보면 객체지향 원칙이 주석이나 설명에도 작동한다는 사실을 알 수 있습니다.

Reply:
아 중요한 걸 놓치고 있었네요..
개발을 할 때만 유연하게가 아니라 말씀하신 주석이나 설명 , 문서 작성에도 **의존**에 대한 고민이 필요하네요
감사합니다 ㅎㅎ